### PR TITLE
XCMv5: Fix for compatibility with V4

### PR DESCRIPTION
**Pull request by yrong**
## Description

Our smoke tests transfer `WETH` from Sepolia to Westend-AssetHub breaks, try to reregister `WETH`  on AH but fails as following:

https://bridgehub-westend.subscan.io/xcm_message/westend-4796d6b3600aca32ef63b9953acf6a456cfd2fbe

https://assethub-westend.subscan.io/extrinsic/9731267-0?event=9731267-2

The reason is that the transact call encoded on BH to register the asset
https://github.com/paritytech/polkadot-sdk/blob/a77940bac783108fcae783c553528c8d5328e5b2/bridges/snowbridge/primitives/router/src/inbound/mod.rs#L282-L289
```
0x3500020209079edaa8020300fff9976782d46cc05630d1f6ebab18b2324d6b1400ce796ae65569a670d0c1cc1ac12515a3ce21b5fbf729d63d7b289baad070139d01000000000000000000000000000000
```

the `asset_id` which is the xcm location can't be decoded on AH in V5

Issue initial post in https://matrix.to/#/!qUtSTcfMJzBdPmpFKa:parity.io/$RNMAxIIOKGtBAqkgwiFuQf4eNaYpmOK-Pfw4d6vv1aU?via=parity.io&via=matrix.org&via=web3.foundation

### CODE REVIEW

**Review by vgeddes — COMMENTED** [review]
Can you be more specific about what the problem is? The element link does not seem to be working for me.

XCMv4 and XCMv5 each have their own `NetworkId` with different encoding layouts, and they need to be explicitly converted to each other using `From` or `TryFrom` traits.

So then surely BH needs to take care to send an XCM to AH that is in the correct format?

File: prdoc/pr_6503.prdoc [hunk 1]
```diff
@@ -0,0 +1,10 @@
+title: "xcm: minor fix for compatibility with V4"
+
+doc:
+  - audience: ["Runtime Dev", "Runtime User"]
+    description: |
+      "To make NetworkId compatibility with V4"
```
Comment (acatangiu) [inline prdoc/pr_6503.prdoc:6]:
```suggestion
      "Following the removal of `NetworkId::Rococo` and `NetworkId::Westend`, fix `xcm::v5::NetworkId` encoding/decoding to be compatible with `xcm::v4::NetworkId`"
```

**Review by acatangiu — APPROVED** [review]

**Comment by acatangiu** [comment]
The PR makes sense if we have v4::NetworkId entries in storage that we want to directly interpret as v5::NetworkId entries without migrating them.

Is this the case?

**Comment by yrong** [comment]
> The PR makes sense if we have v4::NetworkId entries in storage that we want to directly interpret as v5::NetworkId entries without migrating them.
> 
> Is this the case?

Exactly. And I'd assume something wrong or unfinished for the runtime upgrading on BridgeHub. The `SupportedVersion`  is not correct.
image

**Review by alistair-singh — APPROVED** [review]
+1

**Comment by acatangiu** [comment]
cc @franciscoaguirre

**Comment by yrong** [comment]
@franciscoaguirre  Could you help to approve this PR or anything unclear that I can provide, please let me know.

File: prdoc/pr_6503.prdoc [hunk 2]
```diff
@@ -0,0 +1,10 @@
+title: "xcm: minor fix for compatibility with V4"
+
+doc:
+  - audience: ["Runtime Dev", "Runtime User"]
+    description: |
+      "Following the removal of `NetworkId::Rococo` and `NetworkId::Westend`, fix `xcm::v5::NetworkId` encoding/decoding to be compatible with `xcm::v4::NetworkId`"
```
Comment (franciscoaguirre) [inline prdoc/pr_6503.prdoc:6]:
```suggestion
      Following the removal of `Rococo`, `Westend` and `Wococo` from `NetworkId`, fixed `xcm::v5::NetworkId` encoding/decoding to be compatible with `xcm::v4::NetworkId`
```

**Review by franciscoaguirre — APPROVED** [review]
Seems reasonable to keep the same indices

**Comment by acatangiu** [comment]
> @acatangiu For `AssetIdParameter` could we use the `VersionedLocation`
> 
> https://github.com/paritytech/polkadot-sdk/blob/a77940bac783108fcae783c553528c8d5328e5b2/cumulus/parachains/runtimes/assets/asset-hub-westend/src/lib.rs#L417-L418
> instead?

AFAIK this has been discussed before and decided against because the storage still needs an upgrade every time to avoid situations of incorrect duplicate assetids registered: e.g. same ID but once with v3 and once with v4 - in the end it doesn't seem to help and makes code more complicated...

**Comment by yrong** [comment]
@acatangiu I've fixed the conflicts, could you take another look to see whether this PR can be merged?

**Pull request merged**